File: src/app/repair.ts
import { readdir, unlink } from "node:fs/promises";
import { join } from "node:path";
import { ulid } from "ulid";
import { applyEvents } from "../domain/projector";
import { TsqError } from "../errors";
import { appendEvents } from "../store/events";
import { forceRemoveLock, lockExists, withWriteLock } from "../store/lock";
import { getPaths } from "../store/paths";
import { SNAPSHOT_RETAIN_COUNT } from "../store/snapshots";
import type {
  EventPayloadMap,
  EventRecord,
  EventType,
  RelationType,
  RepairPlan,
  RepairResult,
  State,
} from "../types";
import { loadProjectedState, persistProjection } from "./state";

function makeRepairEvent<T extends EventType>(
  actor: string,
  ts: string,
  type: T,
  taskId: string,
  payload: EventPayloadMap[T],
): EventRecord {
  return { event_id: ulid(), ts, actor, type, task_id: taskId, payload: payload as Record<string, unknown> };
}

function buildRepairPlan(
  state: State,
): Omit<RepairPlan, "stale_temps" | "stale_lock" | "old_snapshots"> {
  const orphaned_deps: RepairPlan["orphaned_deps"] = [];
  const orphaned_links: RepairPlan["orphaned_links"] = [];

  for (const [child, blockers] of Object.entries(state.deps)) {
    for (const blocker of blockers) {
      if (!state.tasks[child] || !state.tasks[blocker]) {
        orphaned_deps.push({ child, blocker });
      }
    }
  }

  for (const [src, rels] of Object.entries(state.links)) {
    for (const [kind, targets] of Object.entries(rels)) {
      for (const target of targets ?? []) {
        if (!state.tasks[src] || !state.tasks[target]) {
          orphaned_links.push({ src, dst: target, type: kind as RelationType });
        }
      }
    }
  }

  return { orphaned_deps, orphaned_links };
}

async function scanFilesystem(
  repoRoot: string,
): Promise<{ stale_temps: string[]; stale_lock: boolean; old_snapshots: string[] }> {
  const paths = getPaths(repoRoot);
  const stale_temps: string[] = [];
  let stale_lock = false;
  let old_snapshots: string[] = [];

  try {
    const entries = await readdir(paths.tasqueDir);
    for (const entry of entries) {
      if (entry.includes(".tmp")) {
        stale_temps.push(entry);
      }
    }
  } catch {}

  stale_lock = await lockExists(repoRoot);

  try {
    const snapEntries = await readdir(paths.snapshotsDir);
    const snapshots = snapEntries.filter((name) => name.endsWith(".json")).sort();
    if (snapshots.length > SNAPSHOT_RETAIN_COUNT) {
      old_snapshots = snapshots.slice(0, snapshots.length - SNAPSHOT_RETAIN_COUNT);
    }
  } catch {}

  return { stale_temps, stale_lock, old_snapshots };
}

export async function executeRepair(
  repoRoot: string,
  actor: string,
  now: () => string,
  opts: { fix: boolean; forceUnlock: boolean },
): Promise<RepairResult> {
  if (opts.forceUnlock && !opts.fix) {
    throw new TsqError("VALIDATION_ERROR", "--force-unlock requires --fix", 1);
  }

  const { state } = await loadProjectedState(repoRoot);
  const { orphaned_deps, orphaned_links } = buildRepairPlan(state);
  const { stale_temps, stale_lock, old_snapshots } = await scanFilesystem(repoRoot);
  const plan: RepairPlan = {
    orphaned_deps,
    orphaned_links,
    stale_temps,
    stale_lock,
    old_snapshots,
  };

  if (!opts.fix) {
    return { plan, applied: false, events_appended: 0, files_removed: 0 };
  }

  if (opts.forceUnlock && plan.stale_lock) {
    await forceRemoveLock(repoRoot);
  }

  return withWriteLock(repoRoot, async () => {
    const { state: lockedState, allEvents } = await loadProjectedState(repoRoot);
    const events: EventRecord[] = [];
    const paths = getPaths(repoRoot);

    for (const dep of plan.orphaned_deps) {
      events.push(
        makeRepairEvent(actor, now(), "dep.removed", dep.child, { blocker: dep.blocker }),
      );
    }

    for (const link of plan.orphaned_links) {
      events.push(
        makeRepairEvent(actor, now(), "link.removed", link.src, {
          type: link.type,
          target: link.dst,
        }),
      );
    }

    if (events.length > 0) {
      const nextState = applyEvents(lockedState, events);
      await appendEvents(repoRoot, events);
      await persistProjection(repoRoot, nextState, allEvents.length + events.length);
    }

    let filesRemoved = 0;
    for (const temp of plan.stale_temps) {
      try {
        await unlink(join(paths.tasqueDir, temp));
        filesRemoved++;
      } catch {}
    }

    for (const snap of plan.old_snapshots) {
      try {
        await unlink(join(paths.snapshotsDir, snap));
        filesRemoved++;
      } catch {}
    }

    return { plan, applied: true, events_appended: events.length, files_removed: filesRemoved };
  });
}

File: src/app/repair.rs
use crate::app::state::{load_projected_state, persist_projection};
use crate::domain::deps::normalize_dependency_edges;
use crate::domain::events::make_event;
use crate::domain::projector::apply_events;
use crate::errors::TsqError;
use crate::store::events::append_events;
use crate::store::lock::{force_remove_lock, lock_exists, with_write_lock};
use crate::store::paths::get_paths;
use crate::store::snapshots::SNAPSHOT_RETAIN_COUNT;
use crate::types::{EventRecord, RepairDep, RepairLink, RepairPlan, RepairResult, State};
use serde_json::{Map, Value};
use std::fs::read_dir;
use std::path::Path;

pub struct OrphanedGraph {
    pub orphaned_deps: Vec<RepairDep>,
    pub orphaned_links: Vec<RepairLink>,
}

pub struct RepairOptions {
    pub fix: bool,
    pub force_unlock: bool,
}

pub fn scan_orphaned_graph(state: &State) -> OrphanedGraph {
    let mut orphaned_deps = Vec::new();
    let mut orphaned_links = Vec::new();

    for (child, blockers) in &state.deps {
        for edge in normalize_dependency_edges(Some(blockers)) {
            if !state.tasks.contains_key(child) || !state.tasks.contains_key(&edge.blocker) {
                orphaned_deps.push(RepairDep {
                    child: child.to_string(),
                    blocker: edge.blocker,
                    dep_type: edge.dep_type,
                });
            }
        }
    }

    for (src, rels) in &state.links {
        for (kind, targets) in rels {
            for target in targets {
                if !state.tasks.contains_key(src) || !state.tasks.contains_key(target) {
                    orphaned_links.push(RepairLink {
                        src: src.to_string(),
                        dst: target.to_string(),
                        rel_type: *kind,
                    });
                }
            }
        }
    }

    OrphanedGraph {
        orphaned_deps,
        orphaned_links,
    }
}

fn scan_filesystem(
    repo_root: impl AsRef<Path>,
) -> Result<(Vec<String>, bool, Vec<String>), TsqError> {
    let paths = get_paths(&repo_root);
    let mut stale_temps = Vec::new();
    let mut old_snapshots = Vec::new();

    if let Ok(entries) = read_dir(&paths.tasque_dir) {
        for entry in entries.flatten() {
            if let Some(name) = entry.file_name().to_str()
                && name.contains(".tmp")
            {
                stale_temps.push(name.to_string());
            }
        }
    }

    let stale_lock = lock_exists(repo_root)?;

    if let Ok(entries) = read_dir(&paths.snapshots_dir) {
        let mut snapshots: Vec<String> = entries
            .flatten()
            .filter_map(|entry| entry.file_name().to_str().map(|name| name.to_string()))
            .filter(|name| name.ends_with(".json"))
            .collect();
        snapshots.sort();
        if snapshots.len() > SNAPSHOT_RETAIN_COUNT {
            old_snapshots = snapshots[..snapshots.len() - SNAPSHOT_RETAIN_COUNT].to_vec();
        }
    }

    Ok((stale_temps, stale_lock, old_snapshots))
}

pub fn execute_repair(
    repo_root: impl AsRef<Path>,
    actor: &str,
    now: &dyn Fn() -> String,
    opts: RepairOptions,
) -> Result<RepairResult, TsqError> {
    let repo_root = repo_root.as_ref().to_path_buf();

    if opts.force_unlock && !opts.fix {
        return Err(TsqError::new(
            "VALIDATION_ERROR",
            "--force-unlock requires --fix",
            1,
        ));
    }

    if !opts.fix {
        let loaded = load_projected_state(&repo_root)?;
        let graph = scan_orphaned_graph(&loaded.state);
        let (stale_temps, stale_lock, old_snapshots) = scan_filesystem(&repo_root)?;
        let plan = RepairPlan {
            orphaned_deps: graph.orphaned_deps,
            orphaned_links: graph.orphaned_links,
            stale_temps,
            stale_lock,
            old_snapshots,
        };
        return Ok(RepairResult {
            plan,
            applied: false,
            events_appended: 0,
            files_removed: 0,
        });
    }

    if opts.force_unlock && lock_exists(&repo_root)? {
        let _ = force_remove_lock(&repo_root)?;
    }

    with_write_lock(&repo_root, || {
        let loaded = load_projected_state(&repo_root)?;
        let graph = scan_orphaned_graph(&loaded.state);
        let (stale_temps, stale_lock, old_snapshots) = scan_filesystem(&repo_root)?;
        let plan = RepairPlan {
            orphaned_deps: graph.orphaned_deps,
            orphaned_links: graph.orphaned_links,
            stale_temps,
            stale_lock,
            old_snapshots,
        };

        let mut events: Vec<EventRecord> = Vec::new();

        for dep in &plan.orphaned_deps {
            events.push(make_event(
                actor,
                &now(),
                crate::types::EventType::DepRemoved,
                &dep.child,
                payload_map(serde_json::json!({
                  "blocker": dep.blocker,
                  "dep_type": dep.dep_type,
                })),
            ));
        }

        for link in &plan.orphaned_links {
            events.push(make_event(
                actor,
                &now(),
                crate::types::EventType::LinkRemoved,
                &link.src,
                payload_map(serde_json::json!({
                  "type": link.rel_type,
                  "target": link.dst,
                })),
            ));
        }

        if !events.is_empty() {
            let mut next_state = apply_events(&loaded.state, &events)?;
            append_events(&repo_root, &events)?;
            persist_projection(
                &repo_root,
                &mut next_state,
                loaded.all_events.len() + events.len(),
                None,
            )?;
        }

        let mut files_removed = 0;
        let paths = get_paths(&repo_root);

        for temp in &plan.stale_temps {
            if std::fs::remove_file(paths.tasque_dir.join(temp)).is_ok() {
                files_removed += 1;
            }
        }

        for snap in &plan.old_snapshots {
            if std::fs::remove_file(paths.snapshots_dir.join(snap)).is_ok() {
                files_removed += 1;
            }
        }

        Ok(RepairResult {
            plan,
            applied: true,
            events_appended: events.len(),
            files_removed,
        })
    })
}

fn payload_map(value: Value) -> Map<String, Value> {
    match value.as_object() {
        Some(map) => map.clone(),
        None => Map::new(),
    }
}
